tests/POC/equil_far_opt.py

```diff
@@ -0,0 +1,135 @@
+                        dist = np.sqrt((self.x + dx - other_drone.x)**2 + (self.y + dy - other_drone.y)**2)
+                        cost += (dist - D[drone_index][j])**2
+
+                if cost < min_value:
+                    min_value = cost
+                    best_dir = (dx, dy)
+
+        self.intended_direction = best_dir
+        self.current_cost = min_value
+
+    def move(self):
+        self.x += self.intended_direction[0]
+        self.y += self.intended_direction[1]
+
+def worker(drone, all_drones, idx, D):
+    drone.calculate_direction(all_drones, idx, D)
+    return drone
+
+def individual_drone_score(drone, all_drones, drone_index, D):
+    score = sum([np.sqrt(d.x**2 + d.y**2) for d in all_drones]) / len(all_drones)
+    
+    # Sum of squared differences from the desired distances
+    for j, other_drone in enumerate(all_drones):
+        if j != drone_index:
+            dist = np.sqrt((drone.x - other_drone.x)**2 + (drone.y - other_drone.y)**2)
+            score += (dist - D[drone_index][j])**2
+    
+    return score
+
+
+def individual_drone_score(drone, all_drones, drone_index, D):
+    score = 0
+    
+    # Drone's distance from the origin
+    score += sum([np.sqrt(d.x**2 + d.y**2) for d in all_drones]) / len(all_drones)
+    
+    # Sum of squared differences from the desired distances
+    for j, other_drone in enumerate(all_drones):
+        if j != drone_index:
+            dist = np.sqrt((drone.x - other_drone.x)**2 + (drone.y - other_drone.y)**2)
+            score += (dist - D[drone_index][j])**2
+    
+    return score
+
+
+def simulate_environment(N, D):
+   # Desired center of the equilateral triangle
+    x_center, y_center = 50, 50
+
+    # Shifts based on the triangle's centroid
+    x_shift = x_center - 5
+    y_shift = y_center - (5 * np.sqrt(3) / 3)
+
+    # Updated starting positions as vertices of the translated equilateral triangle
+    optimal_positions = [
+        (0 + x_shift, 0 + y_shift),                
+        (10 + x_shift, 0 + y_shift),               
+        (5 + x_shift, 5 * np.sqrt(3) + y_shift)    
+    ]
+    drones = [Drone(optimal_positions[i]) for i in range(N)]
+    iteration = 0
+
+    positions_history = []
+
+    while True:
+        iteration += 1
+        print(f"Iteration {iteration}: {[(drone.x, drone.y) for drone in drones]}")
+
+        # Print costs of each drone for the current iteration
+        for idx, drone in enumerate(drones):
+            print(f"Cost for Drone {idx}: {drone.current_cost}")
+
+        # Store positions for detecting repetitive patterns
+        current_positions = [(drone.x, drone.y) for drone in drones]
+        positions_history.append(current_positions)
+
+        # Break if the positions repeat
+        if positions_history.count(current_positions) > 2:
+            break
+
+        with concurrent.futures.ThreadPoolExecutor() as executor:
+            drones = list(executor.map(worker, drones, [drones]*N, range(N), [D]*N))
```

Comment:
may want to add a nob to control `max_workers` i.e. 

with `concurrent.futures.ThreadPoolExecutor(max_workers=max_workers)` as executor:

If there's alot of agents/drones, might lead to performance drop depending on user's workstation specs.